server/market/market.go

```diff
@@ -116,35 +121,140 @@ func NewMarket(mktInfo *dex.MarketInfo, storage db.DEXArchivist, swapper Swapper
-	baseCoins, quoteCoins, err := storage.ActiveOrderCoins(mktInfo.Base, mktInfo.Quote)
+	// Load existing book orders from the DB.
+	base, quote := mktInfo.Base, mktInfo.Quote
+	bookOrders, err := storage.BookOrders(base, quote)
 	if err != nil {
 		return nil, err
 	}
-	// TODO: first check that the coins aren't already locked by e.g. another market.
-	log.Debugf("Locking %d base asset coins.", len(baseCoins))
-	for oid, coins := range baseCoins {
-		log.Tracef(" - order %v: %v", oid, coins)
+
+	log.Infof("Loaded %d stored book orders.", len(bookOrders))
+	// Put the book orders in a map so orders that no longer have funding coins
+	// can be removed easily.
+	bookOrdersByID := make(map[order.OrderID]*order.LimitOrder, len(bookOrders))
+	for _, ord := range bookOrders {
+		bookOrdersByID[ord.ID()] = ord
+	}
+
+	baseCoins := make(map[order.OrderID][]order.CoinID)
+	quoteCoins := make(map[order.OrderID][]order.CoinID)
+
+ordersLoop:
+	for id, lo := range bookOrdersByID {
+		if lo.FillAmt > 0 {
+			// Order already matched with another trade, so it is expected that
+			// the funding coins are spent in a swap.
+
+			// In general, our position is that the server is not ultimately
+			// responsible for verifying that all orders have locked coins since
+			// the client will be penalized if they cannot complete the swap.
+			// The least the server can do is ensure funding coins for NEW
+			// orders are unspent and owned by the user, but there are other
+			// possibilities:
+
+			// TODO 1: We may consider following the chain of dex contracts and
+			// redeems to ensure that lo.Remaining() in value is unspent, or at
+			// least check that the initial coins were spent in a dex monitored
+			// tx with TxMonitored(..., spendingTx).
+
+			// TODO 2: The swapper could notify market when a redeem output is
+			// created so it can lock those coins if the order is still on the
+			// book (e.g. signal{orderID, redeemOutCoins} -> market). Otherwise
+			// the user could place new orders with these coins even though they
+			// are implicitly funding the still booked order with a remaining
+			// amount.
+
+			// On to next order. Do not bother locking coins that are spent.
+			continue
+		}
+
+		// Verify all funding coins for this order.
+		assetID := quote
+		if lo.Sell {
+			assetID = base
+		}
+		for i := range lo.Coins {
+			coin, err := swapper.CheckUnspent(assetID, lo.Coins[i])
+			if err == nil {
+				continue
+			}
+
+			if errors.Is(err, asset.CoinNotFoundError) {
+				// spent, exclude this order
+				log.Warnf("Coin %s not unspent for unfilled order %v. "+
+					"Revoking the order.", coin, lo)
+			} else {
+				// other failure (coinID decode, RPC, etc.)
+				log.Errorf("Unexpected error checking coinID %v for order %v: %v. "+
+					"Revoking the order.", lo.Coins[i], lo, err)
+			}
+
+			delete(bookOrdersByID, id)
+			if _, _, err = storage.RevokeOrder(lo); err != nil {
+				log.Errorf("Failed to revoke order %v: %v", lo, err)
+			}
```

Comment:
I think this should only be done if the error was a `CoinNotFoundError`. Otherwise good orders could be set to revoked status because of an rpc connection problem.

Reply:
I'm now having it return an error from constructor in this case since it's likely a configuration or node sync error.